core/monitoring/user-events/src/main/scala/org/apache/openwhisk/core/monitoring/metrics/PrometheusRecorder.scala

```diff
@@ -150,15 +152,111 @@ case class PrometheusRecorder(kamon: PrometheusReporter)
         case ActivationResponse.statusApplicationError => statusApplicationError.inc()
         case ActivationResponse.statusDeveloperError   => statusDeveloperError.inc()
         case ActivationResponse.statusWhiskError       => statusInternalError.inc()
-        case x                                         => statusCounter.labels(namespace, initiator, action, x).inc()
+        case x                                         => promMetrics.statusCounter.labels(namespace, initiator, action, x).inc()
       }
 
       a.size.foreach(responseSize.observe(_))
       a.userDefinedStatusCode.foreach(value =>
-        userDefinedStatusCodeCounter.labels(namespace, initiator, action, value.toString).inc())
+        promMetrics.userDefinedStatusCodeCounter.labels(namespace, initiator, action, value.toString).inc())
     }
   }
 
+  case class PrometheusMetrics() extends PrometheusMetricNames {
+
+    val metricTags: LinkedHashMap[String, String] = LinkedHashMap(
+      actionNamespace -> config.renameTags.getOrElse(actionNamespace, actionNamespace),
+      initiatorNamespace -> config.renameTags.getOrElse(initiatorNamespace, initiatorNamespace),
+      actionName -> config.renameTags.getOrElse(actionName, actionName),
+      actionKind -> config.renameTags.getOrElse(actionKind, actionKind),
+      actionMemory -> config.renameTags.getOrElse(actionMemory, actionMemory),
+      actionStatus -> config.renameTags.getOrElse(actionStatus, actionStatus),
+      userDefinedStatusCode -> config.renameTags.getOrElse(userDefinedStatusCode, userDefinedStatusCode))
+
+    val namespaceActivationTags = metricTags.filterKeys(Set(actionNamespace, initiatorNamespace)).values.toSeq
```

Comment:
This is just a take it or leave and a personal preference but defining a map and then filtering it for extracting the desired values seems like an optimization but it requires more time to understand what is happening there. I would just create the variables and reference them one by one in the counters and histograms as before.